prbt_hardware_support/test/integrationtests/integrationtest_speed_observer_and_operation_mode.cpp

```diff
@@ -0,0 +1,290 @@
+  bool joint_publisher_running_{ false };
+  std_srvs::SetBool::Response sto_res;
+};
+
+void SpeedObserverIntegrationTest::SetUp()
+{
+  subscriber_ =
+      nh_.subscribe<FrameSpeeds>(FRAME_SPEEDS_TOPIC_NAME, 1, &SpeedObserverIntegrationTest::frame_speeds_cb, this);
+  fake_controller_joint_states_pub_ =
+      nh_.advertise<sensor_msgs::JointState>(FAKE_CONTROLLER_JOINT_STATES_TOPIC_NAME, 1);
+  operation_mode_pub_ =
+      nh_.advertise<OperationModes>(OPERATION_MODE_TOPIC, 10);
+
+  pnh_.getParam(ADDITIONAL_FRAMES_PARAM_NAME, additional_frames_);
+  ROS_DEBUG_STREAM("SetUp pnh:" << pnh_.getNamespace() << " nh:" << nh_.getNamespace());
+  for (auto& f : additional_frames_)
+  {
+    ROS_DEBUG_STREAM("- " << f);
+  }
+
+  operation_mode_srv_ =
+      nh_.advertiseService(OPERATION_MODE_SERVICE, &SpeedObserverIntegrationTest::operation_mode_cb, this);
+  sto_srv_ = nh_.advertiseService(STO_SERVICE, &SpeedObserverIntegrationTest::sto_cb, this);
+
+  sto_res.success = true;
+  sto_res.message = "testing ...";
+}
+
+void SpeedObserverIntegrationTest::TearDown()
+{
+  joint_publisher_running_ = false;
+  nh_.shutdown();
+}
+
+void SpeedObserverIntegrationTest::publishOm(OperationModes::_value_type omv){
+  OperationModes om;
+  om.value = omv;
+  om.time_stamp = ros::Time::now();
+  operation_mode_pub_.publish(om);
+}
+
+void SpeedObserverIntegrationTest::publishJointStatesAtSpeed(double v)
+{
+  sensor_msgs::JointState js;
+  js.name = { "testing_world-a" };
+  ros::Rate r = ros::Rate(TEST_FREQUENCY * 3);  // publishing definitely faster
+                                                // then observing
+  ros::Time start = ros::Time::now();
+  double t = 0;
+  joint_publisher_running_ = true;
+  while (joint_publisher_running_ & nh_.ok())
+  {
+    ros::Time current = ros::Time::now();
+    t = (current - start).toSec();
+    js.position = { t * v };
+
+    fake_controller_joint_states_pub_.publish(js);
+    if (joint_publisher_running_ & nh_.ok())  // ending fasteroperation_mode_cb
+      r.sleep();
+  }
+}
+
+void SpeedObserverIntegrationTest::stopJointStatePublisher()
+{
+  joint_publisher_running_ = false;
+}
+
+/**
+ * @tests{Stop1_on_violation_of_speed_limit}
+ *
+ * Test Sequence:
+ *    0. Set Operation Mode to T1 by answering the service call
+ *    1. Joint motion with a speed higher than the T1 limit is faked.
+ *    2. Set Operation Mode to Auto by publihsing on the topic
+ *       Wait for stop calls that may occurr from previous run
+ *    3. Joint motion with a speed lower than the Auto limit is faked.
+ *    4. Switching back to T1
+ *    5. Set up the STO service to return success = false
+ *
+ * Expected Results:
+ *    0. -
+ *    1. Operation mode not to be called any more
+ *       Sto is triggered
+ *    2. -
+ *    3. Operation mode not to be called any more
+ *       Sto is *not* triggered
+ *    4. Sto is triggered
+ *    5. Sto is triggered and error message is produced
+ */
+TEST_F(SpeedObserverIntegrationTest, testT1)
```

Comment:
Correct Test Name? I am missing the test for the additional frame...

Reply:
Done